Sources/ApolloTestSupport/MockGraphQLServer.swift

```diff
@@ -0,0 +1,73 @@
+@testable import Apollo
+import XCTest
+
+public class MockGraphQLServer {
+  enum ServerError: Error, CustomStringConvertible {
+    case unexpectedRequest(String)
+    
+    public var description: String {
+      switch self {
+      case .unexpectedRequest(let requestDescription):
+        return "Mock GraphQL server received an unexpected request: \(requestDescription)"
+      }
+    }
+  }
+  
+  public typealias RequestHandler<Operation: GraphQLOperation> = (HTTPRequest<Operation>) -> JSONObject
+  
+  private class RequestExpectation<Operation: GraphQLOperation>: XCTestExpectation {
+    let file: StaticString
+    let line: UInt
+    let handler: RequestHandler<Operation>
+    
+    init(description: String, file: StaticString = #filePath, line: UInt = #line, handler: @escaping RequestHandler<Operation>) {
+      self.file = file
+      self.line = line
+      self.handler = handler
+
+      super.init(description: description)
+    }
+  }
+  
+  private let queue = DispatchQueue(label: "com.apollographql.MockGraphQLServer")
+    
+  public init() { }
+  
+  // Using existentials to avoid having to deal with type erasure.
+  private var requestExpectations: [AnyHashable: Any] = [:]
+  
+  private subscript<Operation: GraphQLOperation>(_ operationType: Operation.Type) -> RequestExpectation<Operation>? {
+    get {
+      requestExpectations[ObjectIdentifier(operationType)] as! RequestExpectation<Operation>?
+    }
+    
+    set {
+      requestExpectations[ObjectIdentifier(operationType)] = newValue
+    }
+  }
+  
+  public func expect<Operation: GraphQLOperation>(_ operationType: Operation.Type, file: StaticString = #filePath, line: UInt = #line, requestHandler: @escaping (HTTPRequest<Operation>) -> JSONObject) -> XCTestExpectation {
```

Comment:
Do you mean writing JSON in tests using strings instead of dictionaries? I think that would make the tests less readable (no syntax highlighting or automatic formatting), and you wouldn't know you'd written invalid JSON (a missing brace...) until runtime.

Reply:
I do, but in theory once the tests are passing you shouldn't have to futz much with the string itself, so the highlighting/formatting bit is less of a concern to me than "we'd have to rewrite this anyway when we want to change all this to using `Codable` parsing". 

Personally disagree that it makes the tests less readable, especially if you use multi-line string literals, but that may be because I've spent too much of my life staring at raw JSON strings.

Reply:
My worry is that writing JSON by hand without editor support makes it easy to introduce mistakes (like unmatched braces or quotes) you would only catch when running the tests. But you're right that's mostly a one time thing and once the tests are passing it should be ok. 

We don't necessarily have to change this when we switch to `Codable` parsing though. The expectations could still be set up using dictionaries, even if we change the mock server to rendering those as raw JSON when serving requests.

Reply:
Ahhhh i keep forgetting `JSON` is basically a typealias. Maybe we should just skip the `typealias` and have it return the underlying type?

Reply:
Well, the underlying type is `Any` or `[String: Any]` right now, so I think the `typealias` reflects the intention better. My original idea was to eventually make this into a more specific type (like an actual JSON enum) and use `ExpressibleByDictionaryLiteral` etc. to still allow for easy construction during tests.

Reply:
Yeah I've got [something like that in the CodgenLib](https://github.com/apollographql/apollo-ios/blob/main/Sources/ApolloCodegenLib/JSON.swift), though without the `ExpressibleByDictionaryLiteral` support yet